pyproject.toml

```diff
@@ -1,12 +1,68 @@
 [build-system]
-requires = ['setuptools>=40.8.0']
-build-backend = 'setuptools.build_meta'
+requires = ["setuptools>=61.0.0"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "Django"
+dynamic = ["version"]
+requires-python = ">= 3.10"
+dependencies = [
+    "asgiref>=3.7.0",
+    "sqlparse>=0.3.1",
+    "tzdata; sys_platform == 'win32'",
+]
+authors = [
+  {name = "Django Software Foundation", email = "foundation@djangoproject.com"},
+]
+description = "A high-level Python web framework that encourages rapid development and clean, pragmatic design."
+readme = "README.rst"
+license = {text = "BSD-3-Clause"}
+classifiers = [
+    "Development Status :: 2 - Pre-Alpha",
+    "Environment :: Web Environment",
+    "Framework :: Django",
+    "Intended Audience :: Developers",
+    "License :: OSI Approved :: BSD License",
+    "Operating System :: OS Independent",
+    "Programming Language :: Python",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3 :: Only",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
+    "Topic :: Internet :: WWW/HTTP",
+    "Topic :: Internet :: WWW/HTTP :: Dynamic Content",
+    "Topic :: Internet :: WWW/HTTP :: WSGI",
+    "Topic :: Software Development :: Libraries :: Application Frameworks",
+    "Topic :: Software Development :: Libraries :: Python Modules",
+]
+
+[project.optional-dependencies]
+argon2 = ["argon2-cffi>=19.1.0"]
+bcrypt = ["bcrypt"]
+
+[project.scripts]
+django-admin = "django.core.management:execute_from_command_line"
+
+[project.urls]
+Homepage = "https://www.djangoproject.com/"
+Documentation = "https://docs.djangoproject.com/"
+"Release notes" = "https://docs.djangoproject.com/en/stable/releases/"
+Funding = "https://www.djangoproject.com/fundraising/"
+Source = "https://github.com/django/django"
+Tracker = "https://code.djangoproject.com/"
 
 [tool.black]
-target-version = ['py310']
-force-exclude = 'tests/test_runner_apps/tagged/tests_syntax_error.py'
+target-version = ["py310"]
+force-exclude = "tests/test_runner_apps/tagged/tests_syntax_error.py"
 
 [tool.isort]
-profile = 'black'
-default_section = 'THIRDPARTY'
-known_first_party = 'django'
+profile = "black"
+default_section = "THIRDPARTY"
+known_first_party = "django"
+
+[tool.setuptools.dynamic]
+version = {attr = "django.__version__"}
+
+[tool.setuptools.packages]
+find = {}
```

Comment:
So I found that the wheel produced contained `docs/`, `tests/`, etc. that it shouldn't.

It seems to be due to the fact that `setuptools` will consider implicit namespaces by default when using a `pyproject.toml` file according to the note under [this section](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#finding-simple-packages).

We could do the following instead to disable this:

```suggestion
[tool.setuptools.packages.find]
namespaces = false
```

But then I get a large number of warnings in the build output like this:

```
.../lib/python3.12/site-packages/setuptools/command/build_py.py:207: _Warning: Package 'django.conf.app_template' is absent from the `packages` configuration.
!!                                                                                                                      
                                                                                                                        
        ********************************************************************************                                
        ############################                                                                                    
        # Package would be ignored #                                                                                    
        ############################                                                                                    
        Python recognizes 'django.conf.app_template' as an importable package[^1],                                      
        but it is absent from setuptools' `packages` configuration.                                                     
                                                                                                                        
        This leads to an ambiguous overall configuration. If you want to distribute this                                
        package, please make sure that 'django.conf.app_template' is explicitly added                                   
        to the `packages` configuration field.                                                                          
                                                                                                                        
        Alternatively, you can also rely on setuptools' discovery methods                                               
        (for example by using `find_namespace_packages(...)`/`find_namespace:`                                          
        instead of `find_packages(...)`/`find:`).                                                                       
                                                                                                                        
        You can read more about "package discovery" on setuptools documentation page:                                   
                                                                                                                        
        - https://setuptools.pypa.io/en/latest/userguide/package_discovery.html                                         
                                                                                                                        
        If you don't want 'django.conf.app_template' to be distributed and are                                          
        already explicitly excluding 'django.conf.app_template' via                                                     
        `find_namespace_packages(...)/find_namespace` or `find_packages(...)/find`,                                     
        you can try to use `exclude_package_data`, or `include-package-data=False` in                                   
        combination with a more fine grained `package-data` configuration.                                              
                                                                                                                        
        You can read more about "package data files" on setuptools documentation page:                                  
                                                                                                                        
        - https://setuptools.pypa.io/en/latest/userguide/datafiles.html                                                 
                                                                                                                        
                                                                                                                        
        [^1]: For Python, any directory (with suitable naming) can be imported,                                         
              even if it does not contain any `.py` files.                                                              
              On the other hand, currently there is no concept of package data                                          
              directory, all directories are treated like packages.                                                     
        ********************************************************************************                                
                                                                                                                        
!!                                                                                                                      
  check.warn(importable)                                                                                                
```

It turns out we had that before when using `packages = find:` in `setup.cfg` as implicit namespaces weren't considered, but it sounds like something that ought to be fixed. So instead I attempted to explicitly exclude the stuff we don't want in the wheel:

```suggestion
[tool.setuptools.packages.find]
exclude = ["docs*", "extras*", "js_tests*", "scripts*", "tests*"]
```

This works, but I think for safety in case something else is added we should just include exactly what we want:

```suggestion
[tool.setuptools.packages.find]
include = ["django*"]
```

In theory, according to the [flat-layout docs](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#flat-layout) this shouldn't be necessary, as many of these are in the default exclude, but it seems that the `namespaces` thing takes precedence. And we likely want that due to the spewing of warnings that we get.